src/maliput_malidrive/road_curve/spiral_ground_curve.cc

```diff
@@ -0,0 +1,206 @@
+  static constexpr std::array<double, kNumCoefficients> cos_exponents = GetFresnelCosExponents();
+  static constexpr std::array<double, kNumCoefficients> sin_exponents = GetFresnelSinExponents();
+  static constexpr std::array<double, kNumCoefficients> cos_coefficients = GetFresnelCosCoefficients();
+  static constexpr std::array<double, kNumCoefficients> sin_coefficients = GetFresnelSinCoefficients();
+
+  std::array<double, kNumCoefficients> x_terms{};
+  std::array<double, kNumCoefficients> y_terms{};
+  std::fill(x_terms.begin(), x_terms.end(), t);
+  std::fill(y_terms.begin(), y_terms.end(), t);
+
+  for (size_t i = 0; i < kNumCoefficients; ++i) {
+    x_terms[i] = PowerOf(x_terms[i], cos_exponents[i]);
+    y_terms[i] = PowerOf(y_terms[i], sin_exponents[i]);
+  }
+
+  return maliput::math::Vector2{
+      std::inner_product(x_terms.begin(), x_terms.end(), cos_coefficients.begin(), 0.),
+      std::inner_product(y_terms.begin(), y_terms.end(), sin_coefficients.begin(), 0.),
+  };
+}
+
+// @return A vector equivalent of @p v rotated by @p theta.
+inline maliput::math::Vector2 Rotate2dVector(const maliput::math::Vector2& v, double theta) {
+  const double cos_theta = std::cos(theta);
+  const double sin_theta = std::sin(theta);
+  return maliput::math::Vector2(v.x() * cos_theta - v.y() * sin_theta, v.x() * sin_theta + v.y() * cos_theta);
+}
+
+}  // namespace
+
+SpiralGroundCurve::SpiralGroundCurve(double linear_tolerance, const maliput::math::Vector2& xy0, double heading0,
+                                     double curvature0, double curvature1, double arc_length, double p0, double p1)
+    : linear_tolerance_(linear_tolerance),
+      xy0_(xy0),
+      heading0_(heading0),
+      curvature0_(curvature0),
+      curvature1_(curvature1),
+      arc_length_(arc_length),
+      p0_(p0),
+      p1_(p1),
+      validate_p_(maliput::common::RangeValidator::GetAbsoluteEpsilonValidator(p0_, p1_, linear_tolerance_,
+                                                                               GroundCurve::kEpsilon)),
+      k_dot_{(curvature1 - curvature0) / arc_length},
+      norm_{1. / std::sqrt(k_dot_)},
+      spiral_p0_{curvature0 * arc_length / (curvature1 - curvature0)},
+      spiral_heading0_{std::atan2(std::sin(PowerOf(spiral_p0_, 2)), std::cos(PowerOf(spiral_p0_, 2)))},
+      spiral_xy0_{FresnelSpiral(spiral_p0_ / norm_)} {
+  MALIDRIVE_THROW_UNLESS(linear_tolerance_ > 0.);
+  MALIDRIVE_THROW_UNLESS(arc_length_ >= GroundCurve::kEpsilon);
+  MALIDRIVE_THROW_UNLESS(p0_ >= 0.);
+  MALIDRIVE_THROW_UNLESS(p1_ - p0_ >= GroundCurve::kEpsilon);
+  MALIDRIVE_THROW_UNLESS(std::fabs(curvature1_ - curvature0_) >= GroundCurve::kEpsilon);
+  MALIDRIVE_THROW_UNLESS(std::signbit(curvature1_) == std::signbit(curvature0_));
+}
+
+double SpiralGroundCurve::DoPFromP(double xodr_p) const { return validate_p_(xodr_p); }
+
+double SpiralGroundCurve::NormalizedSpiralCoordinateAt(double p) const { return (p - p0_ + spiral_p0_) / norm_; }
```

Comment:
Reminder for me to ask you to walk me through the normalization/denormalization process here.